src/Authoring/WinRT.SourceGenerator/RcwReflectionFallbackGenerator.cs

```diff
@@ -0,0 +1,332 @@
+            executableTypeNames
+            .Where(static names => !names.IsEmpty)
+            .SelectMany(static (executableTypeNames, token) => executableTypeNames.AsImmutableArray())
+            .Collect();
+
+        // Generate the [DynamicDependency] attributes
+        context.RegisterImplementationSourceOutput(projectedTypeNames, static (context, projectedTypeNames) =>
+        {
+            if (projectedTypeNames.IsEmpty)
+            {
+                return;
+            }
+
+            StringBuilder builder = new();
+
+            builder.AppendLine("""
+                // <auto-generated/>
+                #pragma warning disable
+
+                namespace WinRT
+                {
+                    using global::System.Runtime.CompilerServices;
+                    using global::System.Diagnostics.CodeAnalysis;
+
+                    /// <summary>
+                    /// Roots RCW types for assemblies referencing old projections.
+                    /// It is recommended to update those, to get binary size savings.
+                    /// </summary>
+                    internal static class RcwFallbackInitializer
+                    {
+                        /// <summary>
+                        /// Roots all dependent RCW types.
+                        /// </summary>
+                        [ModuleInitializer]                    
+                """);
+
+            foreach (string projectedTypeName in projectedTypeNames)
+            {
+                builder.Append("        [DynamicDependency(DynamicallyAccessedMemberTypes.NonPublicConstructors, typeof(");
+                builder.Append(projectedTypeName);
+                builder.AppendLine("))]");
+            }
+
+            builder.Append("""
+                        public static void InitializeRcwFallback()
+                        {
+                        }
+                    }
+                }
+                """);
+
+            context.AddSource("RcwFallbackInitializer.g.cs", builder.ToString());
+        });
+    }
+
+    /// <summary>
+    /// Checks whether an assembly contains old projections.
+    /// </summary>
+    /// <param name="assemblySymbol">The assembly to inspect.</param>
+    /// <returns>Whether <paramref name="assemblySymbol"/> contains old projections.</returns>
+    private static bool IsOldProjectionAssembly(IAssemblySymbol assemblySymbol)
+    {
+        // We only care about assemblies that have some dependent assemblies
+        if (assemblySymbol.Modules.First() is not { ReferencedAssemblies: { Length: > 0 } dependentAssemblies })
+        {
+            return false;
+        }
+
+        // Scan all dependent assemblies to look for CsWinRT with version < 2.0.8
+        foreach (AssemblyIdentity assemblyIdentity in dependentAssemblies)
+        {
+            if (assemblyIdentity.Name == "WinRT.Runtime")
+            {
+                return assemblyIdentity.Version < new Version(2, 0, 8);
```

Comment:
Given this is an assembly version comparison, we should probably increment the version for this branch to like 2.1.0 until the actual version is determined.

https://github.com/microsoft/CsWinRT/blob/3b983c31ac1889bc78bd45b422bb6b49f0c58ae6/build/AzurePipelineTemplates/CsWinRT-Variables.yml#L5C35-L5C42

Reply:
Checked in a change to address this so that I can also do a Windows SDK projection build to try them together.